### ISSUE TITLE
Add support for Zipkin's ERROR annotation

### ISSUE THREAD

**Issue by jpiechowka** [Nike-Inc/wingtips#21]
I think that adding support for Zipkin's error annotation will be a good addition to Wingtips.

There could be a method like `Tracer.getInstance().completeRequestSpanWithError();` and something similar for sub-spans `Tracer.getInstance().completeSubSpanWithError();`

**Comment by nicmunroe** [comment]
Thanks for the suggestion. Can you provide a link to the Zipkin docs or javadocs or an example where I can see what Zipkin does in these cases?

Also, just a heads up that this may not be possible to support properly until Wingtips has annotation capability for its Spans. That's something I want to add but may not be able to get to for a while.

**Comment by jpiechowka** [comment]
@nicmunroe Here are some docs for Zipkin about the error annotation: 

1. [zipkinCore](https://github.com/openzipkin/zipkin-api/blob/master/thrift/zipkinCore.thrift)
2. [Constants](https://github.com/openzipkin/zipkin/blob/master/zipkin/src/main/java/zipkin/Constants.java)

In Brave it can be done like that for example:

```
SpanId span = brave.localTracer().startNewSpan("Brave", "Example Operation");

brave.localTracer().submitAnnotation("sr");
brave.localTracer().submitAnnotation("error");
brave.localTracer().finishSpan();
```

@adriancole Will know everything how it works.

**Comment by adriancole** [comment]
TL;DR; is..

 adding an annotation of "error" will indicate a potentially transient error
 adding a binary annotation of "error" -> "error message" will mark the
span as failed (you still should add annotations that a response was
received)

**Comment by nicmunroe** [comment]
Thanks for the info. Rather than shoehorn a one-off solution for this into wingtips I think the proper solution will fall out naturally and easily when #23 is done.

**Comment by jpiechowka** [comment]
@nicmunroe I think that makes a lot of sense. When support for custom annotations and binary annotations compatible with Zipkin is added, it will be easy to manually submit an error with a custom message.

**Comment by jpiechowka** [comment]
@adriancole @nicmunroe I was playing today with submitting errors and I came up with a very dirty solution.

In `com.nike.wingtips.zipkin.util.WingtipsToZipkinSpanConverterDefaultImpl.java`

I added:

```java
 public zipkin.Span convertWingtipsSpanToZipkinSpanWithError(Span wingtipsSpan, Endpoint zipkinEndpoint, String localComponentNamespace, String errorMessage) {
        long startEpochMicros = wingtipsSpan.getSpanStartTimeEpochMicros();
        long durationMicros = 0;

        return createNewZipkinSpanBuilderWithError(wingtipsSpan, startEpochMicros, durationMicros, zipkinEndpoint, localComponentNamespace, errorMessage)
                .id(nullSafeLong(wingtipsSpan.getSpanId()))
                .name(wingtipsSpan.getSpanName())
                .parentId(nullSafeLong(wingtipsSpan.getParentSpanId()))
                .timestamp(startEpochMicros)
                .traceId(nullSafeLong(wingtipsSpan.getTraceId()))
                .duration(durationMicros)
                .build();
    }
```

```java
protected zipkin.Span.Builder createNewZipkinSpanBuilderWithErrorBinaryAnnotation(
            Span wingtipsSpan, long startEpochMicros, long durationMicros, Endpoint zipkinEndpoint, String localComponentNamespace, String errorMessage) {
        zipkin.Span.Builder zsb = zipkin.Span.builder();

        switch (wingtipsSpan.getSpanPurpose()) {
            case SERVER:
                zsb.addAnnotation(Annotation.create(startEpochMicros, Constants.SERVER_RECV, zipkinEndpoint))
                        .addAnnotation(Annotation.create(startEpochMicros + durationMicros, Constants.SERVER_SEND, zipkinEndpoint))
                        .addBinaryAnnotation(BinaryAnnotation.create("error", errorMessage, zipkinEndpoint));

                break;
            case CLIENT:
                zsb.addAnnotation(Annotation.create(startEpochMicros, Constants.CLIENT_SEND, zipkinEndpoint))
                        .addAnnotation(Annotation.create(startEpochMicros + durationMicros, Constants.CLIENT_RECV, zipkinEndpoint))
                        .addBinaryAnnotation(BinaryAnnotation.create("error", errorMessage, zipkinEndpoint));

                break;
            case LOCAL_ONLY:
            case UNKNOWN:       // intentional fall-through: local and unknown span purpose are treated the same way
                zsb.addBinaryAnnotation(BinaryAnnotation.create(Constants.LOCAL_COMPONENT, localComponentNamespace, zipkinEndpoint))
                        .addBinaryAnnotation(BinaryAnnotation.create("error", errorMessage, zipkinEndpoint));

                break;
            default:
                logger.warn("Unhandled SpanPurpose type: " + wingtipsSpan.getSpanPurpose().name());
        }

        return zsb;
    }
```

In `com.nike.wingtips.zipkin.util.WingtipsToZipkinSpanConverter.java`

I added this line:
```java
zipkin.Span convertWingtipsSpanToZipkinSpanWithError(Span wingtipsSpan, Endpoint zipkinEndpoint, String localComponentNamespace, String errorMessage);
```

In `com.nike.wingtips.zipkin.WingtipsToZipkinLifecycleListener.java`

I added:

```java
public void spanCompletedWithError(Span span, String errorMessage){
        try {
            zipkin.Span zipkinSpan = zipkinSpanConverter.convertWingtipsSpanToZipkinSpanWithError(span, zipkinEndpoint, localComponentNamespace, errorMessage);
            zipkinSpanSender.handleSpan(zipkinSpan);
        }
        catch(Throwable ex) {

            long currentBadSpanCount = spanHandlingErrorCounter.incrementAndGet();

            // Only log once every MIN_SPAN_HANDLING_ERROR_LOG_INTERVAL_MILLIS time interval to prevent log spam from a malicious (or broken) caller.
            long currentTimeMillis = System.currentTimeMillis();
            long timeSinceLastLogMsgMillis = currentTimeMillis - lastSpanHandlingErrorLogTimeEpochMillis;
            if (timeSinceLastLogMsgMillis >= MIN_SPAN_HANDLING_ERROR_LOG_INTERVAL_MILLIS) {
                // We're not synchronizing the read and write to lastSpanHandlingErrorLogTimeEpochMillis, and that's ok. If we get a few extra
                //      log messages due to a race condition it's not the end of the world - we're still satisfying the goal of not allowing a
                //      malicious caller to endlessly spam the logs.
                lastSpanHandlingErrorLogTimeEpochMillis = currentTimeMillis;

                zipkinConversionOrReportingErrorLogger.warn(
                        "There have been {} spans that were not zipkin compatible, or that experienced an error during span handling. Latest example: "
                                + "wingtips_span_with_error=\"{}\", conversion_or_handling_error=\"{}\"",
                        currentBadSpanCount, span.toKeyValueString(), ex.toString()
                );
            }
        }
    }
```

And after all this I can just write:
```java
WingtipsToZipkinLifecycleListener wingtipsToZipkinLifecycleListener = new WingtipsToZipkinLifecycleListener("Service-Name', "example-local-component-name", "http://localhost:9411");
wingtipsToZipkinLifecycleListener.spanCompletedWithError(Tracer.getInstance().getCurrentSpan(), "Error-Message");
```

**Comment by jpiechowka** [comment]
@adriancole @nicmunroe Using AspectJ and custom annotation I can easily choose what methods I want to trace. Every method is in single sub-span, and when there is exception I can add Zipkin's binary annotation with exception type or exception message.

![wingtips](https://cloud.githubusercontent.com/assets/9040085/18055291/86c7ce76-6e07-11e6-8d69-f562161b3d7c.png)

**Comment by nicmunroe** [comment]
@jpiechowka General tag support was added in version `0.16.0`, and auto-tagging for various server/client wingtips instrumentations (including error tags) was added in version `0.17.0`. So I can finally call this issue resolved. :)

You no longer need to use custom workarounds. Sorry for the long delay!
